Add support for _FILE environment variables

paescuj opened the pull request:
Closes #5895

(`logger.ts` didn't work in `env.ts`, most likely because it was a circular dependency, so I created an additional `logger-no-env.ts`)

Note: Documentation needs be updated accordingly (https://docs.directus.io/reference/environment-variables/)

paescuj committed:
Add support for _FILE environment variables

paescuj committed:
Enhance processing of _FILE env vars

* Same processing as with other env vars (do not simply treat as string)
- tested successfully
* Warn if both variables are set (EXAMPLE and EXAMPLE_FILE)
* Add comments to make it easier to understand the code

rijkvanzanten commented on api/src/env.ts:
```diff
@@ -7,7 +7,7 @@ import dotenv from 'dotenv';
 import { clone, toNumber, toString } from 'lodash';
 import path from 'path';
-import logger from './logger';
+import logger from './logger-no-env';
```
Lets figure out a way we can resolve the circular problem in a way that doesn't require a duplicate of the same logger 🤔 

I think we can resolve it by requiring it later in the flow with an inline require

rijkvanzanten commented on api/src/env.ts:
```diff
@@ -145,7 +145,7 @@ function getEnv() {
 
 	// Default to env vars plain text files
-	return dotenv.parse(fs.readFileSync(configPath).toString());
+	return dotenv.parse(fs.readFileSync(configPath, { encoding: 'utf8' }));
```
Lets await `fs.readFile` instead of using sync. Even though this is once on startup, we shouldn't block the thread

paescuj replied:
Using await `fs.readFile` here would mean that `getEnv` has to be an async function and we would have to await `getEnv` in `let env...`, but this is not supported:
`Top-level 'await' expressions are only allowed when the 'module' option is set to 'esnext' or 'system', and the 'target' option is set to 'es2017' or higher.ts(1378)`

Same for `processValues`.

What should I do? Sorry, unfortunately I'm not an "async" expert...

rijkvanzanten replied:
Hmm yeah good point. That makes this into a way bigger update, as we have to make the environment reading itself async as a whole 🤔 I'll try some things out and report back!

paescuj replied:
Exactly... Thanks! 👍

paescuj replied:
@rijkvanzanten Just let me know if I can support you on this somehow

paescuj replied:
We could maybe get around this by using something like this for now and then switch the environment reading itself to async later on:
```js
env = (async function () {
	await processValues(env);
})();
```

rijkvanzanten replied:
I've been thinking about this quite a bit, and there honestly I don't think there's a good solution. You need the environment inside logger to control it's output (raw vs pretty, what level etc), and you need to be able to log messages in the retrieval of the environment. 🥚 🔁 🐔 

By duplicating the logger instance itself, we ignore the configuration completely, which kinda defeats the purpose of using pino for that particular log, as you can no longer accurately store the output using transformers etc (as it's hardcoded to our defaults at that point). 

We can't really make the env reading async either, for similar reasons. We need all environment variables ready to go before anything else will run, in order to ensure that every package and part of code in the tree can access it immediately. 

Like we discussed in [the other PR](https://github.com/directus/directus/pull/6119), the API should stop booting up when a discrepancy is found in the environment variables, as that will cause unexpected / undesirable behavior, so I think the conclusion here is to use regular console logs, and exit the process when appropriate.

rijkvanzanten commented on api/src/env.ts:
```diff
@@ -175,12 +175,31 @@ function getEnvironmentValueByType(envVariableString: string) {
+		if (key.length > 5 && key.endsWith('_FILE')) {
+			newKey = key.slice(0, -5);
+			try {
+				value = fs.readFileSync(value, { encoding: 'utf8' });
```
Use async

rijkvanzanten commented on api/src/env.ts:
```diff
@@ -175,12 +175,31 @@ function getEnvironmentValueByType(envVariableString: string) {
+		// and store it in the variable with the same name but without '_FILE' at the end
+		let newKey;
+		if (key.length > 5 && key.endsWith('_FILE')) {
+			newKey = key.slice(0, -5);
```
Maybe this could be the point to do a `const logger = require('./logger')`?

paescuj replied:
Unfortunately, this doesn't work either (same error as when using import on top):
`TypeError: logger.warn is not a function`

rijkvanzanten replied:
Oh my bad, it would be `const { default: logger } = require('./logger');` 🤔

paescuj replied:
Hmm, I don't think this is going to work:

```
const { default: logger } = require('./logger.ts');
console.log(logger);
```
Output:
```
undefined
```

Same happens with the import statement:
```
`import logger from './logger';
console.log(logger);
```
Output:
```
undefined
```

I still think this happens because it is a circular dependency (env.ts is getting imported in logger.ts).

paescuj replied:
Workaround in `logger.ts`:
```diff
-import env from './env';
+
+// Prevent circular dependency problem with 'env.ts'
+let env;
+try {
+       env = require('./env');
+} catch {
+       env = {};
+}
```

Would you be happier with this? 😄

rijkvanzanten replied:
Nah, lets keep it simple and go with this: https://github.com/directus/directus/pull/6101#discussion_r648493117

paescuj committed:
Set newKey only after file read was successful

paescuj committed:
Don't convert value > MAX_SAFE_INTEGER to number

Thanks to @skizer!

As stated by @skizer (from #6119):
  Altho it seems that we do have a numerical value
  it can happen that its outside of Number.MAX_SAFE_INTEGER
  thus resulting in a change of the original intended value
  e.g oauth -> discord -> client_id

rijkvanzanten committed:
Merge branch 'main' into support-file-env-vars

rijkvanzanten committed:
Fix recursive logger-env import
